Source/Dafny/Dafny.atg

```diff
@@ -2318,7 +2318,20 @@ FunctionDecl<DeclModifierData dmod, out Function/*!*/ f>
       [ GenericParameters<typeArgs, false> ]
       [ Formals<true, true, isTwoState, true, formals>
       ]
-      [ ":" Type<out returnType> (. SemErr(t, "predicates do not have an explicitly declared return type; it is always bool"); .)
+      [
+        ":"
+        (  IF(FollowedByColon())
+           "("
+             GIdentType<false, false, false, false, out var resultId, out var ty, out var resultIsGhost, out var isOld, out var isNameOnly, out var isOlder>
+             (. Contract.Assert(!resultIsGhost && !isOld && !isNameOnly && !isOlder);
+                if(ty is not BoolType) {
```

Comment:
I understand the more permissive alternative. Here is my reasoning to not allow it (for now)
1) `(result: MyBoolSynonym)`enables strictly more syntax than this PR, so if we decided it's useless, we would not be able to remove it if we added it.
2) I see another benefit using `(result: MyBoolSynonym)`: to use a subset type. But a subset type of bool is not very interesting. If the subset type constrained the value to true, we could simply `ensures result`, and if it contrained the value to false, we could simply `ensures !result`.
3) Despite these workarounds, even if there was some benefit doing `(result: MyBoolSynonym)` (e.g. for special ad-hoc compilation purposes like `type Wrapped<T>=T` that has a special meaning in the target, that could not nicely be implemented with attributes), we could implement it in a follow-up PR after carefully reviewing the need of someone asking for that.
4) Despite this, someone could still convert the predicate to a function for future purposes - this is what people would have done before this PR when trying to name the result.

So, after consideration, I think we can stick to this simpler PR.

Reply:
Sounds good. For now, if someone wants something different than the exact type `bool`, then they can redeclare the predicate as a `function`.

(Btw, there is one other place where Dafny accepts only two very specific types, namely in the optional square brackets that follow an extreme predicate or extreme lemma. The two types accepted are `nat` and `ORDINAL`. I have at times wanted to use a type synonym for them, so that I can easily experiment with changing a whole bunch of declarations from `nat` to `ORDINAL` or vice versa. I hope we'll allow type synonyms there in the future.)